src/hotspot/share/gc/g1/g1FullCollector.hpp

```diff
@@ -71,6 +73,8 @@ class G1FullCollector : StackObj {
   G1FullGCSubjectToDiscoveryClosure _always_subject_to_discovery;
   ReferenceProcessorSubjectToDiscoveryMutator _is_subject_mutator;
 
+  G1FullGCHeapRegionAttr _region_attr_table;
```

Comment:
I don't really see the point of this auxiliary data structure. Why can't we just query the underlying region for its type, pinned, open/close archive?

Reply:
The only reason is performance:

Consider memory access with that auxiliary data structure: it is a single dereference/memory load on a very dense data structure.

Directly querying the HeapRegion* is two dependent dereferences (first getting the HeapRegion*, then to the HeapRegion, then accessing the member) on a fairly large data structure (HeapRegion).

Since these loads are done for every reference (millions of times) the second is much slower, at best filling the cache with useless data as (roughly) a HeapRegion is around a cache line in size (iirc).

Further, with the current encoding of the values in region attribute table, the closed-or-pinned check can be done with a single check instead of some disjunction (ie. region->is_pinned() || region->is_closed(), although it is true that closed regions are always pinned), so saving even more code (and branch locations).

Reply:
I see; thank you for the explanation. Please add some comments around this code.